### ISSUE TITLE
Failure to upload packages

### ISSUE THREAD

**Issue by daviddavis** [pulp/pulp_deb#459]
**Version**
I'm on the main branch (3055a86c41c39acf7c3947b1d274ed0c65ae7fc3).

**Describe the bug**
When I try to upload a package into pulp_deb, it fails.

**To Reproduce**
I've tried with various packages but they all seem to fail. Here's an example using vim:

```
wget http://security.ubuntu.com/ubuntu/pool/main/v/vim/vim_8.2.2434-3ubuntu3_amd64.deb
https --form :/pulp/api/v3/content/deb/packages/ file@vim_8.2.2434-3ubuntu3_amd64.deb
```

**Expected behavior**
Package uploads successfully. 

**Additional context**
I get a failure (`ErrorDetail(string='Unable to read Deb Package', code='invalid')`).

**Comment by daviddavis** [comment]
I think maybe these packages are compressed with zstandard and that's causing the problem?

**Comment by daviddavis** [comment]
Looks like there's a PR upstream to add zstandard support to python-debian:

https://salsa.debian.org/python-debian-team/python-debian/-/merge_requests/79

**Comment by quba42** [comment]
Ugh, why must Ubuntu do things like this?

**Comment by quba42** [comment]
I would expect this to break syncs containing any packages using this compression algorithm as well. Need to verify this!

**Comment by quba42** [comment]
Strangely enough synchronizing `http://security.ubuntu.com/ubuntu/` (impish main amd64 Immediate), which includes the exact package above that fails to upload is successful. I guess during sync the package content is created using the metadata from the relevant Packages file, rather than the package itself.

**Comment by mdellweg** [comment]
> Strangely enough synchronizing `http://security.ubuntu.com/ubuntu/` (impish main amd64 Immediate), which includes the exact package above that fails to upload, is successful. I guess during sync the package content is created using the metadata from the relevant Packages file, rather than the package itself.

Yes indeed. It even creates the Content without ever seeing the real files in case of on on demand sync.

**Comment by daviddavis** [comment]
That was my thought as well since I was surprised that no one else had reported this issue against pulp_deb and ubuntu has been using zstd since impish which was released in October.

> Ugh, why must Ubuntu do things like this?

They really jumped the gun on this. Python for example doesn't support zstd yet, nor does a lot of the tooling like aptly.

**Comment by quba42** [comment]
On a side note, this issue also affects Pulp 2 pulp_deb, also for the package upload. (Does not look like it affects the sync though my test system ran out of disk space before completing it).

**Comment by quba42** [comment]
Another tidbit of information: If you first obtain an affected package via sync, you can then upload it successfully, since the existing package content will be re-used (rather than creating a new one from the uploaded package file).

**Comment by quba42** [comment]
I was able to verify that applying the patch from https://salsa.debian.org/python-debian-team/python-debian/-/merge_requests/79 to python-debian (making sure the zstd binary is available) solves this issue.

I will look into getting in touch with the upstream maintainers to see if they feel like releasing this patch...

**Comment by quba42** [comment]
The upstream MR is now merged to the upstream main branch. Not sure how the `python-debian` release cycle works though.

**Comment by daviddavis** [comment]
Awesome. Are there any changes needed in pulp_deb or will zstd packages just work once python-debian is released?

**Comment by quba42** [comment]
@daviddavis There are no changes needed in pulp_deb, but we should add the `zstd` dependency to the python-debian SPEC file in pulpcore-packaging (I can open a PR there, since I already have the needed commit for our internal packaging).

For any non RPM based Pulp installations this means zstd packages will just work, provided you have the zstd program installed. Now that I think about it, I will re-test if `python-debian` with the patch provides a clear error message when users try to upload a zstd compressed package, but don't have the needed dependency installed. If not, I may open another upstream MR against the library, or alternatively add such error handling to pulp_deb.

**Comment by daviddavis** [comment]
Awesome, let me know if I can help with anything.

**Comment by quba42** [comment]
I am now trying to add slightly improved upstream error messages: https://salsa.debian.org/python-debian-team/python-debian/-/merge_requests/83

**Comment by daviddavis** [comment]
Hey @quba42 looks like there's a new python-debian release:

https://pypi.org/project/python-debian/#history

I retried the reproducer with python-debian 0.1.44 installed but it still seems to be failing (`Unable to read Deb Package`). Is there anything that needs to happen in pulp_deb now?

**Comment by daviddavis** [comment]
Ah, I installed zstd and it worked. I forgot about [your PR](https://github.com/pulp/pulp_deb/pull/513). 😅

I am going to open a PR against the single container to install zstd.

**Comment by daviddavis** [comment]
Single container PR: https://github.com/pulp/pulp-oci-images/pull/208
